lib/class-wp-rest-search-controller.php

```diff
@@ -0,0 +1,832 @@
+ * @since 2.9.0
+ */
+
+/**
+ * Core class to search for posts across multiple post types via the REST API.
+ *
+ * @since 2.9.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Search_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 2.9.0
+	 */
+	public function __construct() {
+		$this->namespace = 'gutenberg/v1';
+		$this->rest_base = 'search';
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes() {
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permission_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to search content.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has search access, WP_Error object otherwise.
+	 */
+	public function get_items_permission_check( $request ) {
+
+		if ( 'edit' === $request['context'] ) {
+			$allowed_post_types = $this->get_allowed_post_types( true );
+
+			if ( empty( $allowed_post_types ) ) {
+				return new WP_Error( 'rest_forbidden_context', __( 'Sorry, you are not allowed to edit content.' ), array( 'status' => rest_authorization_required_code() ) );
+			}
+
+			if ( 'any' !== $request['type'] ) {
+				foreach ( (array) $request['type'] as $post_type ) {
+					if ( 'any' !== $post_type && ! isset( $allowed_post_types[ $post_type ] ) ) {
+
+						/* translators: post type slug */
+						return new WP_Error( 'rest_forbidden_context', sprintf( __( 'Sorry, you are not allowed to edit content of type %s.' ), $post_type ), array( 'status' => rest_authorization_required_code() ) );
+					}
+				}
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves a collection of search results.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+
+		// Ensure a search string is set in case the orderby is set to 'relevance'.
+		if ( ! empty( $request['orderby'] ) && 'relevance' === $request['orderby'] && empty( $request['search'] ) ) {
+			return new WP_Error( 'rest_no_search_term_defined', __( 'You need to define a search term to order by relevance.' ), array( 'status' => 400 ) );
+		}
+
+		// Retrieve the list of registered collection query parameters.
+		$registered = $this->get_collection_params();
+		$query_args = array(
+			'post_status'         => array_keys( $this->get_allowed_post_stati() ),
```

Comment:
`wp_link_query` uses `post_status=>publish`. Is there a reason we're opening things up?

Reply:
Any post status that identifies as its content as public should be searched. That is `publish` only in most cases, but plugins may add to that.

I also included `inherit` to be able to search for attachments. If that should not be possible, I'd be okay with removing that status.

Reply:
We should focus on only supporting `post_status=publish` for now.

Support for `post_status=inherit` opens up a can of permissions-related worms. Custom post statuses should be handled in conjunction with https://github.com/WordPress/gutenberg/issues/3144